server/server_session.cpp

```diff
@@ -6,110 +6,106 @@
+    auto self(shared_from_this());
+    boost::asio::spawn(strand_, [this, self](boost::asio::yield_context yield) {
+        server_processor processor(yield, db_, ss_map_, self);
         try {
-            // control callback is not a completion handler!
-            ws_.control_callback([this](boost::beast::websocket::frame_type kind,
-                                        boost::beast::string_view payload) {
-                // control frames here are pongs
-                boost::ignore_unused(kind, payload);
-                ping_state_ = pingstate::responsive; // responsive
-                timer_.expires_after(std::chrono::seconds(TIMEOUT));
-            });
+            boost::beast::get_lowest_layer(ws_).expires_after(std::chrono::seconds(30));
             ws_.next_layer().async_handshake(ssl::stream_base::server, yield);
+            boost::beast::get_lowest_layer(ws_).expires_never();
+            ws_.set_option(
+                websocket::stream_base::timeout::suggested(boost::beast::role_type::server));
             // Accept the websocket handshake
-            ws_.async_accept_ex(
-                [](boost::beast::websocket::response_type &m) {
-                    // js-websocket(or chrome) require this field to be non-empty
-                    m.insert(boost::beast::http::field::sec_websocket_protocol, "niba-server");
-                },
-                yield);
-
+            ws_.set_option(websocket::stream_base::decorator([](websocket::response_type &res) {
+                // might only need one
+                res.set(http::field::server, "niba-server");
+                res.insert(http::field::sec_websocket_protocol, "niba-server");
+            }));
+            ws_.async_accept(yield);
             for (;;) {
                 // recv request
                 std::string request_str;
                 auto buffer = boost::asio::dynamic_buffer(request_str);
                 nibautil::stopwatch stopwatch_next_request;
-                ws_.async_read(buffer, yield);
+                boost::system::error_code ec;
+                ws_.async_read(buffer, yield[ec]);
+                if (ec) {
+                    BOOST_LOG_SEV(logger_, sev::info) << "Session ending, reason: " << ec.message();
+                    break;
+                }
                 BOOST_LOG_SEV(logger_, sev::info)
-                    << "idled for " << stopwatch_next_request.elapsed_ms() << "ms";                
+                    << "idled for " << stopwatch_next_request.elapsed_ms() << "ms";
                 // check how long the request itself is processed
                 nibautil::stopwatch stopwatch;
-                // reset ping state, and timer as well
-                ping_state_ = pingstate::responsive;
-                timer_.expires_after(std::chrono::seconds(TIMEOUT));
                 // process request and send out response
                 std::string response = processor.dispatch(request_str);
```

Comment:
I know we use non-blocking APIs, but consider if it's still possible for `dispatch` to take a long time (any blocking factor) to return? If so, `write` will be delayed.

Reply:
Yes, that means we took that long to process the user request - they'd have to wait for it